### DISCUSSION TITLE
Fonts used in TKinter - are they packed too?

### DISCUSSION THREAD

**Discussion by aliencaocao** [pyinstaller/pyinstaller#6266] [PyInstaller] [answered]
I have some custom fonts that I downloaded from online and added it into my Windows OS fonts. Now I want to use the custom fonts in my tkinter app and I am not sure if PyInstaller packages these fonts as well so that the same font will be used on other systems without this font installed.
If PyInstaller does not package the font, is there any way for me to manually add it?
The way I'm setting the custom font is this:

```
from tkinter.font import nametofont
tk_default_font = nametofont("TkDefaultFont")  # This is just to modify the default font so that I don't have to set a font attribute for every widget. It is not the font name itself.
tk_default_font.configure(family='custom font name', size=15)  # this is the actual font name installed on my OS.
```

Thanks!

**Comment by rokm** [comment]
PyInstaller does not collect non-python resources as it has no means of recognizing them and keeping track of them. So if your application uses any extra resources, [you need to ensure that they are collected](https://pyinstaller.readthedocs.io/en/stable/spec-files.html#adding-data-files) (for 3rd party packages, existing hooks might already handle their resources).

But in this case, I think you are asking wrong question. Because if the name you pass via `family` argument to `tkinter.font.Font.configure()` method is indeed a font family name, this means that the font needs to be installed in the OS itself. So you'd need to install it on every target computer (or create an installer for your program that also installs the font in the OS). 

If you want to use a font file that's part of your program bundle, then you'll first need to look up how to load a font from a font file into `tkinter` (I don't think `tkinter` itself provides any abstraction for that). And of course ensure that the font file is collected into your frozen application bundle, so your frozen code can load it.

**Comment by bwoodsend** [comment, marked as answer]
An *installed* font is classed as part of the operating system so no, PyInstaller will not collect it. Ideally you want to pass your font to tkinter by its full file path rather than its family name and then just add your font as a regular data file to your application so that neither you or your users have to install anything. TKinter unfortunately seems to lack such a method but, if you don't mind a Windows only solution, you can temporarly install a random font using the Windows C API:
```python
# test.py
import tkinter
import ctypes.wintypes
from pathlib import Path

wingdi = ctypes.CDLL("gdi32")
wingdi.AddFontResourceExW.argtypes = [ctypes.c_wchar_p, ctypes.wintypes.DWORD, ctypes.c_void_p]


def add_font(font_path):
    """Temporarily install a new font."""
    return wingdi.AddFontResourceExW(str(font_path), 0x10, 0)
   

def main():
    """
    An example program which loads a custom Explora-Regular.ttf font, taken from
    https://fonts.google.com/specimen/Explora and placed in the same folder as
    this script, then uses it in a tkinter label.
    
    """
    add_font(Path(__file__).with_name("Explora-Regular.ttf"))
   
    root = tkinter.Tk()
    root.geometry("500x200")
    label = tkinter.Label(root, text="Hello in some fancy font", font=('Explora', 35))
    label.pack()
    root.mainloop()
    
    
if __name__ == "__main__":
    main()
```
Then on the PyInstaller end, it would just be:
```
pyinstaller --add-data Explora-Regular.ttf;. test.py
```

**Reply by aliencaocao** [reply]
Thanks for the detailed answer. I've got two questions. Does this font stay in the target system, or will it be removed after the app exits? And does this operation require admin rights? Thanks!

**Reply by bwoodsend** [reply]
No, that's what the that's what the `0x10` flag (stands for [FR_PRIVATE](https://docs.microsoft.com/en-us/windows/win32/api/wingdi/nf-wingdi-addfontresourceexa#parameters)) does. The font is available only to this processes, is removed when the program terminates and doesn't require escalated privileges.
